KZL-298-allow-user-to-created-computed-fields

etrousset opened the pull request:
## What does this PR do ?
First version of the computed field plugin for Kuzzle
Allow user to create field in document that are computed using other fields of the same document

### How should this be manually tested?

Download, install, read the doc, run the tests, and have fun with it.

etrousset committed:
Merge pull request #8 from kuzzleio/kzl-51-add-automated-test

Add end2end test

Aschen committed:
Change docker image

Aschen committed:
Merge pull request #9 from kuzzleio/change-docker-image

Change docker image

Aschen committed:
Bind node inpector on 0.0.0.0

Njuelle committed:
Merge pull request #11 from kuzzleio/add-node-inspector

Bind node inspector on 0.0.0.0

etrousset committed:
more implementation

etrousset committed:
recompute: added action to recalculate computed fields

-this need to be called after one or several computed fields
template has changed

etrousset committed:
gitignore

etrousset committed:
recompute: removed document replace return result from

recompute result

etrousset committed:
document update: added get missing fields

-when updating a document, if the update doesn't contain all
the source fields needed to calculate a computed field, then the
document is get to retreive the missing source fields

etrousset committed:
functional tests: computed fields management

etrousset committed:
functional tests: added tests for computed fields

etrousset committed:
functional tests: more tests and bug fixes

etrousset committed:
functionnal test: fixed a timeout in BeforeAll step

-cumcumber was making step et connection to kuzzle fail due to the BeforeAll
timedout silently...

etrousset committed:
code cleanup

etrousset committed:
dev-stack: add launch script for plugin dev

-the script will set a proper PLUGIN_NAME from the package.json
file

etrousset committed:
computed field confguration: removed the need for sourceFields

-removed the need to provide sourceFields list. It is now extracted
from the value template

etrousset committed:
computedFields: removed sourceFields field from storage...

etrousset committed:
http route: fixed to be more Kuzzle style

etrousset committed:
doc: first draft

etrousset committed:
doc: more doc

etrousset committed:
doc: more doc

Aschen commented:
As we discussed, rather than using eval, here is a solution with only regexp and replace: https://gist.github.com/Aschen/c070ad4ad14a8437b7ab1149c06fc86f

etrousset committed:
tests: made tests happier

etrousset committed:
Merge branch 'KZL-298-allow-user-to-created-computed-fields' of github.com:kuzzleio/computef-fields-plugin into KZL-298-allow-user-to-created-computed-fields

alexandrebouthinon committed:
Add PR reporting

kuzzle commented:
<h1>:cucumber: : Cucumber Report</h1><p>  <em>Check this <a href='https://dl.kuzzle.io/tests-report/computed-fields-plugin/1/cucumber-report.html'>link</a></em></p>

alexandrebouthinon committed:
Fix kuttlefish

etrousset committed:
test: fixed test script wouldn't return non null value when tests faild

alexandrebouthinon committed:
Update comment template

etrousset committed:
test: added 'wait_for' when recomputing computed fields

alexandrebouthinon committed:
Update template

alexandrebouthinon committed:
Update Travis run

etrousset committed:
test: fix fix fix

etrousset committed:
computed-fileds: removed debug log

etrousset committed:
test & bug fixes party:

-got rid of eval()
-fixed all inappropriate usage of var
-fixed scrolling through document would enter the recursion even if not needed
-fixed various bugus bugs...

etrousset committed:
features and tests: more of both

-Added, recomputing computed fields for a collection now remove computed field that
have been deleted
-added test for delete computed fields (shall not be added any more to new documents,
and should be deleted when a recompute is performed)
-fixed World constructor to be able to use cucumber attach() method
-attached some step info to help troubleshooting tests without
having to put console.logs everywhere...

etrousset commented:
> As we discussed, rather than using eval, here is a solution with only regexp and replace: https://gist.github.com/Aschen/c070ad4ad14a8437b7ab1149c06fc86f
 @Aschen : Done

scottinet commented on .eslintrc:
```diff
@@ -54,5 +53,12 @@
     "es6": true
   },
+  "parserOptions": {
+    "ecmaVersion": 8,
```
Kuzzle supports Node.js 6 and upper. Official plugins should follow these requirements, since they are executed by the same node.js than Kuzzle.

Therefore, you must stick to ES6.

scottinet commented on README.md:
```diff
@@ -1,75 +1,262 @@
+## Purpose
 
-This plugin is useful only if you use it as the starting point of your work. It's a boilerplate.
+The purpose of this plugin is to allow one to create fields that are computed from other fields of a document. e.g. considering the following document:
```
Proposal: `This plugin allows creating computed fields from a document's existing fields.  For instance, considering the following document:`

scottinet commented on README.md:
````diff
@@ -1,75 +1,262 @@
+}
+```
+
+you could have a computed field `fullname` defined as `"${name} ${surname}"` and whose value would be `"Michael Romeo"` for previous exemple.
````
-`for previous exemple`

scottinet commented on README.md:
```diff
@@ -1,75 +1,262 @@
+
+you could have a computed field `fullname` defined as `"${name} ${surname}"` and whose value would be `"Michael Romeo"` for previous exemple.
+
+The computed fileds are ensted in a `_computedFields` object to avoid collision with document fields, so the final document will be:
```
ensted?

Do you mean `saved`, or `stored` perhaps?

etrousset replied:
I meant `nested` but `stored` could do also?

scottinet replied:
`embedded` would be closer to what you originally meant, otherwise `stored` is fine

scottinet commented on README.md:
```diff
@@ -1,75 +1,262 @@
+
+| Action      | Description                                                     |
+| ----------- | --------------------------------------------------------------- |
+| `create`    | Adds or update a computed field to an `index/collection`        |
```
Add

scottinet commented on README.md:
```diff
@@ -1,75 +1,262 @@
+
+**http**
+
+You can create or update a computed field using a http `POST` on the route: `http+https://kuzzlehost:kuzzleport/computed-fields/`
```
on the route => using the route

scottinet commented on README.md:
```diff
@@ -1,75 +1,262 @@
+
+You can create or update a computed field using a http `POST` on the route: `http+https://kuzzlehost:kuzzleport/computed-fields/`
+
+For exemple:
```
example

scottinet commented on README.md:
```diff
@@ -1,75 +1,262 @@
-$ docker-compose -f docker/docker-compose.yml up
+### List computed fields
+
+The `list` action will return a array of computed fields.
```
an array

scottinet commented on README.md:
````diff
@@ -1,75 +1,262 @@
+
+```javascript
+kuzzle
+  .query({
````
:warning: The current JS SDK is v5, and in that version, `query` does not return a promise: it takes a callback instead

ballinette replied:
... or call `queryPromise` ?

scottinet commented on README.md:
```diff
@@ -1,75 +1,262 @@
+
+**http**
+
+Do a `POST` on route `http+https://kuzzlehost:kuzzleport/:_id/_delete'` to delete computed field with `_id`
```
Shouldn't this be a `DELETE` route instead?

Also, the route is incorrect, as the provided ID is not attached to any index or collection arguments. I did see that the generated ID depends on these two arguments, but I find this system fragile.

The route should be: 

`DELETE http+https://kuzzlehost:kuzzleport/_plugin/computed-fields/:index/:collection/:_id`

or:
`DELETE http://<host>:<port>/_plugin/computed-fields/<index>/<collection>/<_id>`

scottinet commented on README.md:
```diff
@@ -1,75 +1,262 @@
+**http**
 
-#### Customizing the plugin instance name
+Do a `GET` on route `http+https://kuzzlehost:kuzzleport/computed-fields/` to get computed field list
```
Wrong route, it should be: `http+https://kuzzlehost:kuzzleport/_plugin/computed-fields`

Also, I'm quite fond of the standard `<param>` to describe a required parameter. What do you think of describing your routes like that: `http://<host>:<port>/_plugin/computed-fields`

Lastly, I'm uneasy at the possibility given by that plugin to retrieve any and all computed fields, regardless of their index/collection arguments. This bypasses our right system, and this is also against what a HTTP `GET` route should be, meaning a resource retrieval.
To me, the route should be defined like this: `GET http://<host>:<port>/_plugin/computed-fields/<index>/<collection>`

scottinet commented on README.md:
````diff
@@ -1,75 +1,262 @@
 
-You can use the `docker-compose.yml` file included in this repository to start a development-oriented stack to help you creating your custom Kuzzle Core plugin.
+```sh
+$ curl  localhost:7512/_plugin/computed-fields/ -H "Content-Type: application/json"
````
This route is wrong: the body part of an HTTP request should only contain... well... body arguments. And neither `index` nor `collection` are body arguments.

The route should be more like this:

`POST http://<host>:<port>/_plugin/computed-fields/<index>/<collection>`

Or even:

`PUT http://<host>:<port>/_plugin/computed-fields/<index>/<collection>/<name>`
With only the `value` argument remaining in the body part of the request.

scottinet commented on docker/docker-compose.yml:
```diff
@@ -2,7 +2,7 @@ version: '3'
 services:
   kuzzle:
-    image: kuzzleio/kuzzle:${KUZZLE_DOCKER_TAG:-latest}
+    image: kuzzleio/development:${KUZZLE_DOCKER_TAG:-latest}
```
plugin-dev?

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+    this.storage = this.context.accessors.storage
+    this.computedFields = []
+    this.sourceFields = {}
+    this.initStorage()
```
`initStorage` should be returned to allow Kuzzle to wait for its completion

Also, the `catch` part is unnecessary if you return the initStorage promise: Kuzzle already provides logs if it fails, and it will stop its starting sequence. With your code, Kuzzle will continue to function, which does not seem to be a good thing.

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+    return this.computedFieldsRepository.search({})
+      .then(r => {
+        r.hits.forEach(e => {
+          delete e[Symbol.for('_kuzzle')]
```
?

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
-    Promise.resolve(/* any value */);
+  computeField(document, sourceFields, valueExpression) {
+    let value = valueExpression
+    sourceFields.forEach((field) => {
```
unnecessary parens around `field`

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+      }
+      else
+      {
+        throw new Error(`Document ${JSON.stringify(document)} doesn't contain field '${fieldname}'`)
```
Throwing a plain Error will make Kuzzle recast it as a `PluginImplementationError`, but it seems more like a `BadRequestError` to me.
Also, I don't know if this isn't better to simply ignore a missing field instead.

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+    const fields = fieldname.split('.')
+    let cum = ''
+    let value = document
+    for (let field of fields) {
```
nitpicking: make `field` const.

As this is a nitpicking remark, I won't report any other occurences of it for the remaining of this PR.

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
-    // check credentials existence
-    Promise.resolve(/* true|false */);
+  getSourceFields(template) {
+    let reg=/\$\{([\w_$][\w\d_$.]+)\}/g
```
should be const

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
-    Promise.resolve(/* true|false */);
+  getSourceFields(template) {
+    let reg=/\$\{([\w_$][\w\d_$.]+)\}/g
+    let sf = []
```
also const

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+    for(let cf of cfs) {
+      if (collection === cf.collection && index === cf.index) {
+
+        if (typeof request.input.body._computedFields === 'undefined') {
```
If `_computedFields` is defined and not an object, then the code below will make this plugin crash, forcing Kuzzle to log a `PluginImplementationError` error

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+              })
+              return this.context.accessors.execute(getReq)
+                .then(res => {
+                  const document = {...res.result._source, ...request.input.body}
```
This doesn't work in node.js 6

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+      }
+    }
+
+    return Promise.all(promises)
```
The promise should not be returned if it ends up resolving a callback

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+  createComputedField (request) {
+    if (request.input.body 
+      && request.input.body.name
+      && request.input.body.index
```
Inconsistent with Kuzzle API: `index` and `collection` are made available at the following path: `request.input.resource.index` and `request.input.resource.collection`, respectively

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+          return this.loadComputedFields().then(() => res) 
+        })
+    }
+    return Promise.reject('Body must contain the following fields: \'name\', \'index\', \'collection\' and \'value\'')
```
Use template litterals so you don't have to escape single quotes

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+  }
+
+  deleteComputedField (request) {
+    let _id = request.input.args._id || request.input.resource._id
```
should be const

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+      remaining,
+      total
+/*
+    if (cfs.length === 0) {
```
Commented out code should be removed

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
-      });
-    }
+  recomputeCollectionComputedFields(request, index, collection) {
+    const cfs = this.getCollectionConputedFields(index, collection)
```
missing semicolon

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+    let searchRequest,
+      scrollId,
+      remaining,
+      total
```
missing semicolon

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+        from: 0,
+        size: this.config.page_size
+      }
+    )
```
missing semicolon

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
-      message: 'Login failed - Reason: You shall not pass!'
-    });
+    return this.context.accessors.execute(searchRequest)
+      .then((res) => {
```
unnecessary parens around `res`

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
-    });
+    return this.context.accessors.execute(searchRequest)
+      .then((res) => {
+        let documents = res.result.hits.map(e => {return {_id: e._id, body: e._source}})
```
* should be const
* missing semicolon

Also, you can return an object directly from within an arrow function, using the following syntax:

`const documents = res.result.hits.map(e => ({_id: e._id, body: e_source}}));`

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+    return this.context.accessors.execute(searchRequest)
+      .then((res) => {
+        let documents = res.result.hits.map(e => {return {_id: e._id, body: e._source}})
+        scrollId = res.result._scroll_id
```
missing semicolon

I just noticed that there almost aren't any semicolon in this section of code, so I'll stop reporting this (but they still must be fixed)

I recommend adding eslint using Kuzzle's `.eslintrc.json` configuration file to this project

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+        remaining = total - res.result.hits.length
+        return this.recomputeFieldsForDocs(documents, index, collection, cfs)
+      })
+      .then((res) => {
```
unnecessary parens

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+          return this.processNextScroll(scrollId, remaining, index, collection, cfs)
+        }
+      })
+      .then((res) => {
```
ditto

scottinet commented on lib/index.js:
```diff
@@ -184,283 +53,409 @@ class CorePlugin {
+        }
+      })
+      .then((res) => {
+        return Promise.resolve({updatedDocuments: total})
```
Inside a promise chain, everything you return is chainable. Thus you don't need to return a new promise, you can simply return `return {updatedDocuments: total};`

scottinet commented on report/index.js:
```diff
@@ -0,0 +1,23 @@
+var reporter = require('cucumber-html-reporter');
```
const

scottinet commented on report/index.js:
```diff
@@ -0,0 +1,23 @@
+var reporter = require('cucumber-html-reporter');
+const {Kuzzle} = require('kuzzle-sdk')
+
+var options = {
```
const

scottinet commented on report/index.js:
```diff
@@ -0,0 +1,23 @@
+const kuzzle = new Kuzzle('websocket', {host:'localhost', port: 7512})
+kuzzle.connect()
+  .then(() => kuzzle.server.info())
+  .then((info) => {
```
unnecessary parens

scottinet reviewed (changes requested):
I still question the usefulness of this plugin, considering that you can do the exact same thing with a much simpler plugin, adding a pipe to API routes that write documents.

ballinette commented on package.json:
```diff
@@ -1,37 +1,39 @@
+    "cucumber-html-reporter": "^4.0.4",
     "eslint": "^3.6.1",
     "istanbul": "^0.4.5",
+    "kuzzle-sdk": "git+https://github.com/kuzzleio/sdk-javascript#6-dev",
```
nitpicking: use `6-beta` version to stay in a more stable release

ballinette replied:
but why do you instantiate your own SDK instance and not use the embedded one that has been introduced in `1.5.0` release (see https://github.com/kuzzleio/kuzzle/pull/1179) ?

etrousset replied:
Because I started working on it before this feature was released

ballinette commented on lib/index.js:
```diff
@@ -1,29 +1,7 @@
- *
- * @see http://docs.kuzzle.io/plugins-reference/plugins-creation-prerequisites/
- */
 class CorePlugin {
```
nitpicking: we could rename the plugin class with a more explicit name (`ComputedFiledsPlugin` for instance ?)

etrousset committed:
huge update: made changes according to PR feedback

-Computed fields are now handle on per collection basis
-Delete is now more restfull: using DELETE verb on <index>/<collection>/<computedFiledName>
-Documentation update
...

etrousset committed:
code cleanup: removed commented code

etrousset committed:
tests: added test for admin controller

etrousset committed:
test: added HTTP routes testing

etrousset committed:
doc update

etrousset committed:
rename plugins class

etrousset committed:
fix tests

etrousset committed:
fixed: plugin was still using kuzzle development docker image

etrousset committed:
removed commented code

scottinet commented on features/support/steps.js:
```diff
@@ -1,19 +1,26 @@
 'use strict'
-var _ = require('lodash');
+var _ = require('lodash')
```
* should be `const`
* missing semicolon

scottinet commented on features/support/steps.js:
```diff
@@ -1,19 +1,26 @@
+  When,
+  Then
+} = require('cucumber'),
+  rp = require('request-promise')
```
Trailing semicolons are missing from this entire file.

scottinet commented on features/support/steps.js:
```diff
@@ -22,49 +29,104 @@ Given('a collection {string} in {string}', function (collectionName, indexName)
+Given('I create a new computed field in index {string} and collection {string} using HTTP API as follow:',
+  function (index, collection, computedFieldCfg) {
+    computedFieldCfg = JSON.parse(computedFieldCfg)
+    let cfName = computedFieldCfg.name
```
should be `const`

scottinet commented on features/support/steps.js:
```diff
@@ -22,49 +29,104 @@ Given('a collection {string} in {string}', function (collectionName, indexName)
+})
+
+Then('the list contains computed field', function (computedField) {
+  let cf = this.result.find((e) => _.isMatch(e, JSON.parse(computedField)))
```
should be `const`

scottinet commented on features/support/steps.js:
```diff
@@ -73,18 +135,30 @@ Then('the list doesn\'t contain computed field {string}', function (cfName) {
-        this.computedFieldIDs = {...this.computedFieldIDs, ...{[cfName]: res.result._id}}
-        this.attach(JSON.stringify(res, undefined, 2))
-      })
+    let cfBody = {
```
should be `const`

scottinet commented on lib/index.js:
```diff
@@ -289,31 +274,42 @@ class CorePlugin {
-    let _id = request.input.args._id || request.input.resource._id
-    if (_id) {
-      return this.removeComputedField(_id)
+    let index = request.input.resource.index
```
should be `const`

Alternatively, you can use destructuring syntax: 
`const {index, collection} = request.input.resource;`

scottinet commented on lib/index.js:
```diff
@@ -289,31 +274,42 @@ class CorePlugin {
-      return this.removeComputedField(_id)
+    let index = request.input.resource.index
+    let collection = request.input.resource.collection
+    let name = request.input.body && request.input.body.name || request.input.args.name
```
should be `const`

scottinet commented on lib/index.js:
```diff
@@ -289,31 +274,42 @@ class CorePlugin {
 
   listComputedFields (request) {
-    return Promise.resolve(this.computedFields)
+    let index = request.input.resource.index
```
see above

scottinet commented on lib/index.js:
```diff
@@ -289,31 +274,42 @@ class CorePlugin {
+    let collection = request.input.resource.collection
+    let result = this.computedFields.filter(e=> e.index === index && e.collection === collection)
+    result = result.map(e => {
+      let elem = JSON.parse(JSON.stringify(e))
```
should be `const`

scottinet commented on lib/index.js:
```diff
@@ -450,12 +442,19 @@ class CorePlugin {
+    return this.computedFieldsRepository.search({}, 0, 1000)
+    .then(r => {
+      let p = []
+      r.hits.forEach((e)=> p.push(this.computedFieldsRepository.delete(e._id)))
```
* `p` should be `const`
* unnecessary parens around `e`
* can be simplified with a map:
`const p = r.hits.map(e => this.computedFieldsRepository.delete(e._id));`

scottinet commented:
Also, the file `lib/index.js` still needs trailing semicolons added.

I recommend you add the following rule to your eslint configuration file: `"semi": [2, "always"]`

etrousset committed:
applied resquested changes

etrousset committed:
added semicolons in cucumber steps

etrousset committed:
adding forgoten test scenarios

scottinet reviewed (approved):
:+1:
